Ported signing enclaves to Windows.

gupta-ak opened the pull request.

gupta-ak commented:
bors try

gupta-ak commented:
bors try

anakrish commented on 3rdparty/mbedtls/CMakeLists.txt:
```diff
@@ -60,3 +61,36 @@ target_link_libraries(mbedcrypto INTERFACE
+install (FILES ${OE_LIBDIR}/openenclave/host/${MBEDTLS_LIBRARY_NAME} DESTINATION ${CMAKE_INSTALL_LIBDIR}/openenclave/host)
+
+ExternalProject_Get_Property(mbedtls-wrap-host BINARY_DIR)
+add_library(mbedcryptohost INTERFACE)
```
Instead of creating an mbedtls library for host, does it make sense to just included the relevant mbedtls files in the CMakeLists.txt for host.
Currently it gives the impression that oe host requires multiple crypto libraries, whereas we just need a fraction of mbedtls capabilities.

anakrish commented on host/sgxsign.c:
```diff
@@ -71,50 +79,238 @@ static oe_result_t _get_date(unsigned int* date)
+     * which is normally calculated as S^3 mod M. We see that we can
+     * derive these Q1 & Q2 values:
+     *  - S^3 mod M
+     *    -> (S * (S^2 mod M)) mod M
```
(a * b) mod n = ((a mod n) * (b mod n)) mod n
With a = S, b = S^2, and n = M
shouldn't this be: ((S modM) * (S^2 mod M)) mod M?

gupta-ak replied:
Yeah, I skipped a step here:

The property is that `(a * b) mod n = (( a mod n) * (b * mod n)) mod n`. To calculate the RSA signature `S`, you do `S = (plaintext)^d mod n` , so we have `S < n` and consequently, `S = (S mod n)`.

anakrish commented on host/sgxsign.c:
```diff
@@ -71,50 +79,238 @@ static oe_result_t _get_date(unsigned int* date)
+     *  - S^3 mod M
+     *    -> (S * (S^2 mod M)) mod M
+     *  - S^2 mod M
+     *    -> S^2 = FLOOR(S^2 / M) * M + R1
```
Can either of S and M be negative numbers? If so the formula can be tricky depending on what mod means.

-17 mod 5 is 3 (ie -17 = 4 * -5 + 3) according to mathematical minded languages like MATLAB, Octave etc.
They have two functions mod and rem that capture the nuance. You can try octave here: https://octave-online.net/

-17 mod 5 is -2 (ie -17 = 3 * -5 + -2) according to common programming languages like C.

So what kind of mod does crypto want?

gupta-ak replied:
They can't be negative. M is the modulus, so it must be positive. S is created by taking the mod, so it will be `0 <= S < M` . I think the mbedtls `mbedtls_mpi_read_binary ` function will assume that the number is unsigned. 

The mod that crypto wants is the mathematical definition. So -17 mod 5 should be 3. I think the mbedtls does this but it doesn't specify. It states that `P / Q` will give `P = Q * N + R`.

CodeMonkeyLeet commented on host/crypto/windows/rsa.c:
```diff
@@ -0,0 +1,556 @@
+        oe_memcpy_s(
+            buffer,
+            *buffer_size,
+            keybuf + sizeof(*keyblob) + keyblob->cbPublicExp,
```
Probably worth a comment to describe the `BCRYPT_RSAKEY_BLOB` as being the header concatenated with the public exp and modulus in that order, here and in `oe_rsa_public_key_get_exponent`.

gupta-ak replied:
sounds good.

gupta-ak commented on 3rdparty/mbedtls/CMakeLists.txt:
```diff
@@ -60,3 +61,36 @@ target_link_libraries(mbedcrypto INTERFACE
+else()
+# On Windows, we compile the mbedcrypto library so we can use it for the bignum functions.
+
+if (CMAKE_GENERATOR MATCHES "Visual Studio")
```
Visual studio generates the binaries in `.../[Debug|Release]/...` unlike the other generators like Ninja or NMake, it's specific to the vs generator. I will add a comment.

gupta-ak commented on host/crypto/windows/rsa.c:
```diff
@@ -0,0 +1,556 @@
+#include <wincrypt.h>
+
+/*
+ * Note that these structures were copied from the crypto/keys.c file.
```
sounds good.

gupta-ak commented on host/crypto/windows/rsa.c:
```diff
@@ -0,0 +1,556 @@
+        ULONG sig_size;
+        NTSTATUS status;
+        BCRYPT_PKCS1_PADDING_INFO info;
+        uint8_t hash_data_copy[64];
```
The one in the outside is not used and should be removed.

gupta-ak commented on host/crypto/windows/rsa.c:
```diff
@@ -0,0 +1,556 @@
+{
+    oe_result_t result = OE_UNEXPECTED;
+    const oe_private_key_t* impl = (const oe_private_key_t*)private_key;
+    uint8_t* hash_data_copy = NULL;
```
this one is not used and should be removed.

gupta-ak commented on host/CMakeLists.txt:
```diff
@@ -12,7 +12,7 @@ set(PLATFORM_SRC
     crypto/key.c
     crypto/random.c
-    crypto/rsa.c
+    crypto/linux/rsa.c
```
sounds good.

ihaller replied:
Why linux and not openssl? Or is the host always going to use openssl on Linux? crypto/openssl/rsa.c makes more sense in a pluggable crypto model. Similarly cypto/bcrypt/rsa.c.

gupta-ak replied:
I changed to crypto/openssl and crypto/bcrypt.

gupta-ak commented on host/sgxload.c:
```diff
@@ -737,7 +737,7 @@ oe_result_t oe_sgx_initialize_enclave(
                 sizeof(info.EInitToken),
                 (void*)&launch_token,
-                sizeof(launch_token)));
+                sizeof(info.EInitToken)));
```
Hm what do you mean? The original code would error and prevent you from initializing enclaves since it was copying `sizeof(launch_token)`, but the static assert only checked if `sizeof(info.EnitToken) <= sizeof(launch_token)`. So, you only want to copy `sizeof(info.EInitToken)`
from `launch_token`, which was my change.

gupta-ak commented on host/crypto/windows/rsa.c:
```diff
@@ -0,0 +1,556 @@
+        switch (hash_type)
+        {
+            case OE_HASH_TYPE_SHA256:
+                if (hash_size > 32)
```
yeah that check makes more sense.

gupta-ak commented on host/crypto/windows/rsa.c:
```diff
@@ -0,0 +1,556 @@
+{
+    oe_result_t result = OE_UNEXPECTED;
+    const oe_private_key_t* impl = (const oe_private_key_t*)private_key;
+    uint8_t* hash_data_copy = NULL;
```
yeah should be removed.

gupta-ak commented on host/crypto/windows/rsa.c:
```diff
@@ -0,0 +1,556 @@
+    OE_CHECK(_get_public_rsa_blob_info(impl->pkey, &keybuf, &keybuf_size));
+
+    /* Sanity check to ensure we get modulus and public exponent. */
+    key_header = (BCRYPT_RSAKEY_BLOB*)keybuf;
```
sounds good.

gupta-ak commented on host/crypto/windows/rsa.c:
```diff
@@ -0,0 +1,556 @@
+// Licensed under the MIT License.
+
+#include "../rsa.h"
+#define WIN32_NO_STATUS
```
If I don't have the `WIN32_NO_STATUS` macro, then there's a lot of macro redef warnings no matter how I rearrange those headers. The redefs have due to ntstatus.h and um/winnt.h containing the same defs. You still need ntstatus.h though since STATUS_SUCCESS isn't defined.

gupta-ak commented on host/sgxsign.c:
```diff
@@ -71,50 +79,238 @@ static oe_result_t _get_date(unsigned int* date)
+    return result;
+}
+
+#if defined(_WIN32)
```
Yeah, I'll write up an issue with future crypto refactor work for windows.

gupta-ak commented on 3rdparty/mbedtls/CMakeLists.txt:
```diff
@@ -60,3 +61,36 @@ target_link_libraries(mbedcrypto INTERFACE
 	${OE_LIBDIR}/openenclave/enclave/liboecore.a)
+
+else()
+# On Windows, we compile the mbedcrypto library so we can use it for the bignum functions.
```
as discussed yesterday, I changed this to just include the bignum file in the host.

gupta-ak commented on host/crypto/windows/rsa.c:
```diff
@@ -0,0 +1,556 @@
+} oe_public_key_t;
+
+/* Magic numbers for the RSA key implementation structures */
+static const uint64_t _PRIVATE_KEY_MAGIC = 0x7bf635929a714b2c;
```
I'm going to defer this to the future crypto refactor, since we did have some discussion if these should be the same for Windows / Linux.

gupta-ak commented on host/crypto/windows/rsa.c:
```diff
@@ -0,0 +1,556 @@
+OE_STATIC_ASSERT(sizeof(oe_public_key_t) <= sizeof(oe_rsa_public_key_t));
+OE_STATIC_ASSERT(sizeof(oe_private_key_t) <= sizeof(oe_rsa_private_key_t));
+
+bool oe_private_key_is_valid(const oe_private_key_t* impl)
```
yeah

gupta-ak commented on host/crypto/rsa.h:
```diff
@@ -5,9 +5,19 @@
+#if !defined(_WIN32)
 /* Caller is responsible for validating parameters */
 void oe_rsa_public_key_init(oe_rsa_public_key_t* public_key, EVP_PKEY* pkey);
+#endif
```
Yeah I'll write an issue for it.

gupta-ak commented:
bors try

gupta-ak commented:
bors try

gupta-ak commented:
@CodeMonkeyLeet @anakrish I updated the PR based off your feedback.

I moved all the openssl code to `host/crypto/openssl` and the bcrypt code to `host/crypto/bcrypt`. The only things left are the `rsa.h` header, which is platform indepdendent and the `sha.c` which has `#ifdef` for openssl/bcrypt (for the major crypto refactor we should split this into 2 files).

gupta-ak commented:
bors try

CodeMonkeyLeet commented on host/crypto/openssl/rsa.h:
```diff
@@ -0,0 +1,13 @@
+/* Caller is responsible for validating parameters */
+void oe_rsa_public_key_init(oe_rsa_public_key_t* public_key, EVP_PKEY* pkey);
+
+#endif /* _OE_HOST_CRYPTO_OPENSSL_RSA_H */
```
Fix missing newline

gupta-ak commented:
bors r+
